ntp-os-clock/src/lib.rs

```diff
@@ -98,109 +110,103 @@ fn duration_in_nanos(duration: NtpDuration) -> libc::c_long {
-                // We have nanosecond precision. use it
-                ntp_kapi_timex.time.tv_usec as u32
-            } else {
-                (ntp_kapi_timex.time.tv_usec as u32) * 1000
-            },
-        ))
+        Ok(extract_current_time(&ntp_kapi_timex))
     }
 
-    fn set_freq(&self, freq: f64) -> Result<(), Self::Error> {
+    fn set_frequency(&self, freq: f64) -> Result<NtpTimestamp, Self::Error> {
         let mut ntp_kapi_timex = EMPTY_TIMEX;
         ntp_kapi_timex.modes = libc::MOD_FREQUENCY;
         // NTP Kapi expects frequency adjustment in units of 2^-16 ppm
         // but our input is in units of seconds drift per second, so convert.
         ntp_kapi_timex.freq = (freq * 65536e6) as libc::c_long;
-        if unsafe { libc::ntp_adjtime(&mut ntp_kapi_timex as *mut _) } != -1 {
-            // We don't care here about the time status, so the non-error
-            // information in the return value of ntp_adjtime can be ignored
-            Ok(())
-        } else {
-            Err(convert_errno())
-        }
+        adjtime(&mut ntp_kapi_timex)?;
+        Ok(extract_current_time(&ntp_kapi_timex))
     }
 
-    fn step_clock(&self, offset: ntp_proto::NtpDuration) -> Result<(), Self::Error> {
-        let mut tp = libc::timespec {
-            tv_sec: 0,
-            tv_nsec: 0,
-        };
-
-        let (offset_secs, offset_nanos) = offset.as_seconds_nanos();
-
-        // Begin time critical section
-        // any time spend between here and the clock_settime call will reduce the
-        // accuracy of the made step
-        if unsafe { libc::clock_gettime(libc::CLOCK_REALTIME, &mut tp as *mut _) } == -1 {
-            return Err(convert_errno());
-        }
-
-        tp.tv_sec += offset_secs as libc::time_t;
-        tp.tv_nsec += offset_nanos as libc::c_long;
-        if tp.tv_nsec >= 1_000_000_000 {
-            // Deal with carry from addition of nanosecond parts
-            tp.tv_nsec -= 1_000_000_000;
-            tp.tv_sec += 1;
-        }
+    fn step_clock(&self, offset: ntp_proto::NtpDuration) -> Result<NtpTimestamp, Self::Error> {
+        let mut timex = EMPTY_TIMEX;
+        timex.modes = libc::ADJ_SETOFFSET | libc::MOD_NANO;
+        let (secs, nanos) = offset.as_seconds_nanos();
+        timex.time.tv_sec = secs as libc::time_t;
+        timex.time.tv_usec = nanos as libc::suseconds_t;
+        adjtime(&mut timex)?;
+        Ok(extract_current_time(&timex))
+    }
 
-        if unsafe { libc::clock_settime(libc::CLOCK_REALTIME, &tp as *const _) } == -1 {
-            return Err(convert_errno());
-        }
-        // End time critical section
+    fn enable_ntp_algorithm(&self) -> Result<(), Self::Error> {
+        let mut timex = EMPTY_TIMEX;
+        adjtime(&mut timex)?;
+        timex.modes = libc::MOD_STATUS;
+        timex.status |= libc::STA_PLL;
+        timex.status &= !libc::STA_FLL & !libc::STA_PPSTIME & !libc::STA_PPSFREQ;
```

Comment:
could `!a & !b = !(a | b)` be used to clean this up? maybe some comments are nice here too, something like "this turns off/on the flags for ... "

Reply:
Done